clang/lib/CIR/Dialect/Transforms/LoweringPrepare.cpp

```diff
@@ -28,16 +65,149 @@ struct LoweringPreparePass : public LoweringPrepareBase<LoweringPreparePass> {
 
   /// Tracks current module.
   ModuleOp theModule;
+
+  /// Tracks existing dynamic initializers.
+  llvm::StringMap<uint32_t> dynamicInitializerNames;
+  llvm::SmallVector<FuncOp, 4> dynamicInitializers;
 };
 } // namespace
 
+cir::FuncOp LoweringPreparePass::buildCXXGlobalVarDeclInitFunc(GlobalOp op) {
+  SmallString<256> fnName;
+  {
+    std::unique_ptr<clang::MangleContext> MangleCtx(
+        astCtx->createMangleContext());
+    llvm::raw_svector_ostream Out(fnName);
+    auto varDecl = op.getAst()->getAstDecl();
+    MangleCtx->mangleDynamicInitializer(varDecl, Out);
+    // Name numbering
+    uint32_t cnt = dynamicInitializerNames[fnName]++;
+    if (cnt)
+      fnName += "." + llvm::Twine(cnt).str();
+  }
+
+  // Create a variable initialization function.
+  mlir::OpBuilder builder(&getContext());
+  builder.setInsertionPointAfter(op);
+  auto fnType = mlir::cir::FuncType::get(
+      {}, mlir::cir::VoidType::get(builder.getContext()));
+  FuncOp f = builder.create<mlir::cir::FuncOp>(op.getLoc(), fnName, fnType);
+  f.setLinkageAttr(mlir::cir::GlobalLinkageKindAttr::get(
+      builder.getContext(), mlir::cir::GlobalLinkageKind::InternalLinkage));
+  mlir::SymbolTable::setSymbolVisibility(
+      f, mlir::SymbolTable::Visibility::Private);
+  mlir::NamedAttrList attrs;
+  f.setExtraAttrsAttr(mlir::cir::ExtraFuncAttributesAttr::get(
+      builder.getContext(), attrs.getDictionary(builder.getContext())));
+
+  // move over the initialzation code of the ctor region.
+  auto &block = op.getCtorRegion().front();
+  mlir::Block *EntryBB = f.addEntryBlock();
+  EntryBB->getOperations().splice(EntryBB->begin(), block.getOperations(),
+                                  block.begin(), std::prev(block.end()));
+
+  // Replace cir.yield with cir.return
+  builder.setInsertionPointToEnd(EntryBB);
+  auto &yieldOp = block.getOperations().back();
+  assert(isa<YieldOp>(yieldOp));
+  builder.create<ReturnOp>(yieldOp.getLoc());
+  return f;
+}
+
+void LoweringPreparePass::lowerGlobalOp(GlobalOp op) {
+  auto &ctorRegion = op.getCtorRegion();
+  if (!ctorRegion.empty()) {
+    if (!ctorRegion.hasOneBlock()) {
+      op.emitError() << "ctor region must have exactly one block.";
+      return;
+    }
+    auto &block = ctorRegion.front();
+    if (block.empty()) {
+      op.emitError() << "ctor region shall not be empty.";
```

Comment:
Same for this one, if a empty ctor region isn't a desired state, should also be part of GlobalOp verifier.